sycl/doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc

```diff
@@ -0,0 +1,200 @@
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Introduction
+
+This extension adds functionality to convert value of single-precision
+floating-point type(`float`) to `bfloat16` type and vice versa. The extension
+doesn't add support for `bfloat16` type as such, instead it uses 16-bit integer
+type(`uint16_t`) as a storage for `bfloat16` values.
+
+The purpose of conversion from float to bfloat16 is to reduce ammount of memory
+required to store floating-point numbers. Computations are expected to be done with
+32-bit floating-point values.
+
+
+== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an implementation
+supporting this extension must predefine the macro
+`SYCL_EXT_INTEL_BF16_CONVERSION` to one of the values defined in the table
+below. Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro’s
+ value to determine which of the extension’s APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+== Extension to `enum class aspect`
+
+[source]
+----
+namespace sycl {
+enum class aspect {
+  ...
+  ext_intel_bf16_conversion
+}
+}
+----
+
+If a SYCL device has the `ext_intel_bf16_conversion` aspect, then it natively
+supports conversion of values of `float` type to `bfloat16` and back.
+
+If the device doesn't have the aspect, objects of `bfloat16` class ust not be
+used in the device code.
+
+== New `bfloat16` class
+
+The following class provides the conversion functionality:
+
+[source]
+----
+namespace sycl {
+namespace ext {
+namespace intel {
+namespace experimental {
+
+class [[sycl_detail::uses_aspects(ext_intel_bf16_conversion)]]
+bfloat16 {
+  using storage_t = uint16_t;
+  storage_t value;
+
+public:
+  // Direct initialization
+  bfloat16(const storage_t& a)
+
+  // Convert from float to bfloat16
+  bfloat16(const float& a);
+
+  // Convert from bfloat16 to float
+  operator float() const;
+
+  // Get bfloat16 as uint16.
+  operator storage_t() const;
+};
+
+} // namespace experimental
+} // namespace intel
+} // namespace ext
+} // namespace sycl
+----
+
+== Example
+
+[source]
+----
+bfloat16 operator+(const bfloat16 &lhs, const bfloat16 &rhs) {
+  return static_cast<float>(lhs) + static_cast<float>(rhs);
+}
+
+float foo(float a, float b) {
+  // Convert from float to bfloat16.
+  bfloat16 A {a};
+  bfloat16 B {b};
+
+  // Convert A and B from bfloat16 to float, do addition on floating-pointer
+  // numbers, then convert the result to bfloat16 and store in C.
+  bfloat16 C = A + B;
```

Comment:
I believe, as is the example won't be compiled. Are these A and B conversions are being done implicitly or users have to add explicit cast call? I believe, we can overload math operators in BFloat16 class to make them casting right and left operands to float and then process the math operation itself. Otherwise please add `static_cast<float>` here.

Reply:
Please note `operator+` overloading 10 lines above

Reply:
Missed it, thanks. But what about adding the overload to the class directly in both spec and implementation https://github.com/intel/llvm/pull/4213/files ? It does look like to be a reasonable enhancement.

Reply:
Added operator overloadings in a7354b4